### DISCUSSION TITLE
Migrate scripts to xtask pattern

### DISCUSSION THREAD

**Discussion by simonbuchan** [dudykr/stc#1096] [Ideas]
As a filthy Windows user, the current scripts setup for tests etc., is a bit fiddly to work with.

I have a branch that has an initial stab at adding the xtask pattern and migrating `crates/stc_ts_type_checker/scripts/check.sh` to it, which seems to have worked out pretty well - the code is quite straightforward and the only real downside is it's a crate that has to be built the first time, and `Command` is a bit uglier to work with directly than shell. On the other side, the Rust to filter, trim and sort the output of the conformance tests pass list is quite simple, in comparison to shell.

Check out the branch here: https://github.com/dudykr/stc/compare/main...simonbuchan:stc:xtask

For reference, the xtask pattern is essentially the closest equivalent to npm scripts for cargo: see [this blog post](https://betterprogramming.pub/running-rust-tasks-with-xtask-and-xtaskops-a2193e67dc25) for a description; but the TDLR is you can just clone and run `cargo xtask check` on any platform instead of `crates/stc_ts_type_checker/scripts/check.sh` on linux / mac.

Is this something you would be interested in?

**Comment by kdy1** [comment]
I think it's a good idea

**Comment by simonbuchan** [comment]
Neat!

The main ones to start with are probably the ones in https://stc.dudy.dev/docs/contributing - I'm thinking starting with something like:

```
cargo xtask [command]

COMMANDS:

  test-analyzer   replaces crates/stc_ts_file_analyzer test scripts
    --fast        replaces ./scripts/fast.sh
    --all         replaces ./scripts/base.sh with no args
    <NAME>        replaces ./scripts/base.sh with a test name (but allows partial matches?)

  auto-unignore   replaces crates/stc_ts_file_analyzer/scripts/auto-unignore.sh

  test-checker    replaces crates/stc_ts_type_checker/scripts/check.sh
     <NAME>       replaces crates/stc_ts_type_checker/scripts/test.sh <NAME>
```

I'm happy to iterate on workflow if you have any ideas, though!

**Comment by simonbuchan** [comment]
Opened a draft in #1097 but seems I can't mark PRs as draft :(

There's a few question mark details there about what variables should be set, etc., but the API above is basically there.

- The `--features tracing/max_level_*` flags caused a lot of rebuilding, so I'm not putting them in there for now since the `RUST_LOG` value is set whenever it is used - I'm not sure if it's behaving as expected?
- There's a `touch .../tests/base.rs` in there that needs a rust nightly feature that looks like it will stabilize soon: https://github.com/rust-lang/rust/issues/98245 - and I dunno, seems like it works!? Not sure what it did.
